### ISSUE TITLE
Return from finally block ?

### ISSUE THREAD

**Issue by apblack** [gracelang/language#24]
Should we allow a __return__ from the finally block in a try() catch() .... finally() expression?  If so, what does it mean?

Javascript allows this, and says that the value of the whole expression is the value returned form the finally block, regardless of whether one of the other blocks might already have executed a return.  What do other languages do?

At present, with minigrace, return is allowed from the finally block when using the JS backend — the effect is to return the value from the finally.  With the C backend, we get a segfault.

**Comment by kjx** [comment]
> On 31/08/2015, at 18:55pm, Andrew Black <notifications@github.com> wrote:
> 
> Should we allow a return from the finally block in a try() catch() .... finally() expression?

how could we prevent it?  dynamically? 

> If so, what does it mean?

return immediately with that value.

method foo {                                                            
  try {4}                                                                
   catch { _ -> print "throwing muses" }                               
   finally { return 42 }                                               
}  

in kernan this returns 42

J

**Comment by apblack** [comment]
I think that we agreed to follow the path of least resistance, and allow "early termination" of a finally clause, either with a `return`, or by raising an exception.  In such a situation, any prior `return` or raised exception is silently dropped.

**Comment by kjx** [comment]
drafted in at - https://github.com/gracelang/language/blob/portland-james/spec.md#catching-exceptions

**Comment by KimBruce** [comment]
Here is a variant of the earlier discussion that regards the semantics of finally.  Suppose we are given the following methods:
```
method tryFinallyNoReturn  {
    try {
       4
    } finally {
        "OK"
    }
}

method tryFinallyReturn {
    try {
        return 4
    } finally {
        "OK"
    }
}
```
The question is what is returned by each of those methods?  The spec makes clear that a finally with a return statement overrides a return in either try or catch clauses, but doesn't say what happens when a value is expected and the finally clause contains a value.

In minigrace, both the above methods return 4.  That makes perfect sense to me in the second method, but is less intuitive in the first.  I don't find returning "OK" to be that much more compelling, but think that we need something in the language spec about it.  If we prefer the minigrace answer then perhaps adding something to the effect that finally clauses are treated as statements and only return a value if they contain an explicit return statement.

**Comment by apblack** [comment]
Yes, we should add to the spec something along the lines of "finally clauses are treated as statements and only return a value if they contain an explicit return statement."

**Comment by apblack** [comment]
Please review the wording in the [above-mentioned pull-request](https://github.com/gracelang/language/pull/179).
